hs-abci-sdk/src/Tendermint/SDK/BaseApp/Metrics/Prometheus.hs

```diff
@@ -0,0 +1,205 @@
+                                                                Sem, interpretH,
+                                                                pureT, raise,
+                                                                runT)
+import           Polysemy.Reader                               (Reader (..),
+                                                                ask)
+import qualified System.Metrics.Prometheus.Concurrent.Registry as Registry
+import qualified System.Metrics.Prometheus.Http.Scrape         as Http
+import qualified System.Metrics.Prometheus.Metric.Counter      as Counter
+import qualified System.Metrics.Prometheus.Metric.Histogram    as Histogram
+import qualified System.Metrics.Prometheus.MetricId            as MetricId
+import           Tendermint.SDK.BaseApp.Metrics                (CountName (..), HistogramName (..),
+                                                                Metrics (..))
+
+--------------------------------------------------------------------------------
+-- eval
+--------------------------------------------------------------------------------
+
+evalWithMetrics
+  :: Member (Embed IO) r
+  => Member (Reader (Maybe MetricsConfig)) r
+  => Sem (Metrics ': r) a
+  -> Sem r a
+evalWithMetrics action = do
+  mCfg <- ask
+  case mCfg of
+    Nothing  -> evalNothing action
+    Just cfg -> evalMetrics (metricsState cfg) action
+
+evalNothing
+  :: Sem (Metrics ': r) a
+  -> Sem r a
+evalNothing = do
+  interpretH (\case
+    IncCount _ -> pureT ()
+    WithTimer _ _ -> pureT ()
+    )
+
+evalMetrics
+  :: Member (Embed IO) r
+  => MetricsState
+  -> Sem (Metrics ': r) a
+  -> Sem r a
+evalMetrics state@MetricsState{..} = do
+  interpretH (\case
+    -- | Increments existing count; if it doesn't exist, creates a new
+    -- | counter and increments it.
+    IncCount ctrName -> do
+      let c@MetricIdentifier{..} = countToIdentifier ctrName
+          cid = metricIdStorable c
+          cMetricIdName = MetricId.Name metricIdName
+      liftIO $ modifyMVar_ metricsCounters $ \counterMap ->
+        case Map.lookup cid counterMap of
+          Nothing -> do
+            newCtr <- liftIO $
+              Registry.registerCounter cMetricIdName metricIdLabels metricsRegistry
+            let newCounterMap = insert cid newCtr counterMap
+            liftIO $ Counter.inc newCtr
+            pure newCounterMap
+          Just ctr -> do
+            liftIO $ Counter.inc ctr
+            pure counterMap
+      pureT ()
+
+    -- | Updates a histogram with the time it takes to do an action
+    -- | If histogram doesn't exist, creates a new one and observes it.
+    WithTimer histName action -> do
+      start <- liftIO $ getCurrentTime
+      a <- runT action
+      end <- liftIO $ getCurrentTime
+      let time = fromRational . (* 1000.0) . toRational $ (end `diffUTCTime` start)
+      observeHistogram state histName time
+      _ <- raise $ evalMetrics state a
```

Comment:
I never gotten it to typecheck otherwise. The result of `evalMetrics state a` (at least for our cases) is almost always `()`, but I don't think the type checker knows that given the type of the `withTimer` summand is over a general `m a`.